### ISSUE TITLE
[Question] Should async-compression behave in a way it does?

### ISSUE THREAD

**Issue by Andrey36652** [Nullus157/async-compression#315]
Hello, I'm trying to fix this bug https://github.com/seanmonstar/reqwest/issues/2381 in reqwest library. I have found out a reason and (preliminary) a way to fix it.
But I'm not sure if this is a bug in async-compression itself or in a way reqwest uses it.
Could you take a look and clarify?

**Comment by NobodyXu** [comment]
> Seems like I've found out a reason. async-compression by default will report stream as "Done" as soon as GzipDecoder has finished it's work.

I think that's the right behavior.

And

> But one can change this behaviour by calling decoder.multiple_members(true) after constructing GzipDecoder.

It's also right.

I believe that reqwest shouldn't test eof by checking our decoder, because eof in our code is used to know if the archive has been consumed.

I think you would want to get back the inner connection and test it directly?

**Comment by Andrey36652** [comment]
> to get back the inner connection and test it directly

How can I do this? I'm new to rust and these libs so `decoder.multiple_members(true)` was my best suggestion. And from the way the code in `decoder.rs` is structured I guessed that layered structure is author's intention.

**Comment by NobodyXu** [comment]
There is `get_ref()`, `get_mut()` and `into_inner()` that enables you to get the inner I/O object from the decoder.

**Comment by Andrey36652** [comment]
Is this okay https://github.com/seanmonstar/reqwest/commit/c9faaebc7d05ea5b27d19e94865ebfd2b29cfe78 ?
It passes my tests https://github.com/Andrey36652/reqwest/tree/inner-connection-eof-solution, but I'm not sure if Pin shenanigans are implemented properly.

**Comment by NobodyXu** [comment]
Yes it's correct, since no unsafe is used and the code compiles, your Pin code is working as intended.

`Pin::new` is for objects that do not contain any self-ref.

It detects that through the `Unpin` trait, any type implements type means that there won't be any self-ref in the object created.

**Comment by Andrey36652** [comment]
Thank you, it's done now.
